dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,385 @@
+import math
+import numpy as np
+import scipy.stats as stats
+from functools import partial
+from pprint import pprint
+from toolz import reduce
+import toolz
+import warnings
+from timeit import default_timer
+
+import sklearn
+from dask_ml.model_selection._split import train_test_split
+
+from sklearn import clone
+from sklearn.model_selection import ParameterSampler
+
+import dask.array as da
+from .model_selection import DaskBaseSearchCV, GridSearchCV, _RETURN_TRAIN_SCORE_DEFAULT
+
+
+def _train(model, X, y, X_val, y_val, max_iter=1, dry_run=False, classes=None,
+           s=1, i=1, k='_', verbose=False, **fit_kwargs):
+    """
+    Train function. Returns validation loss.
+
+    Takes max_iters as parameters. THese are treated as a "scarce resource" --
+    Hyperband is trying to minimize the total iters use for max_iters.
+    """
+    start_time = default_timer()
+    for iter in range(int(max_iter)):
+        if verbose:
+            msg = ("Training model {k} in bracket s={s} iteration {i}. "
+                   "This model is {percent:.1f}% trained for this bracket")
+            print(msg.format(iter=iter, k=k, s=s, i=i, percent=iter * 100.0 / max_iter))
+        if not dry_run:
+            _ = model.partial_fit(X, y, classes, **fit_kwargs)
+    fit_time = default_timer() - start_time
+
+    start_time = default_timer()
+    score = model.score(X_val, y_val) if not dry_run else np.random.rand()
+    score_time = default_timer() - start_time
+    return score, {'score_time': score_time, 'fit_time': fit_time}
+
+
+def _top_k(choose_from, eval_with, k=1):
+    assert k > 0, "k must be positive"
+    assert set(choose_from.keys()) == set(eval_with.keys())
+
+    keys = list(choose_from.keys())
+    assert k <= len(keys)
+
+    eval_with = [eval_with[key] for key in keys]
+    idx = np.argsort(eval_with)
+
+    return {keys[i]: choose_from[keys[i]] for i in idx[-k:]}
+
+
+def _random_choice(params, n=1):
+    configs = ParameterSampler(params, n)
+    return [{k: v if not isinstance(v, np.ndarray) else da.from_array(v)
+            for k, v in config.items()} for config in configs]
+
+
+def _successive_halving(params, model, n=None, r=None, s=None, verbose=False,
+                        shared=None, eta=3, _prefix='', dry_run=False, **fit_kwargs):
+    client = _get_client()
```

Comment:
We'll want to

- document that this uses the concurrent.futures API, and so requires an active `Client`
- Raise a nice error message when one isn't present
- test that it raises